File: Tools/Annotation/Labeling/api_handling/api_wrapper/points_and_joints.py
import requests
import traceback

from api_wrapper.main import base_url, auth

def get_points_for_project(project_id):
    try:
        response = requests.get(f"{base_url()}/get_points_for_project", json={"project_id": project_id}, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def get_joints_for_project(project_id):
    try:
        response = requests.get(f"{base_url()}/get_joints_for_project", json={"project_id": project_id}, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def create_point(project_id, name, color = "white"):
    try:
        data = {"project_id": project_id, "name": name, "color": color}
        response = requests.post(f"{base_url()}/create_point", json=data, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def rename_point(point_id, new_name):
    try:
        data = {"point_id": point_id, "new_name": new_name}
        response = requests.post(f"{base_url()}/rename_point", json=data, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def delete_point(point_id):
    try:
        response = requests.post(f"{base_url()}/delete_point", json={"point_id": point_id}, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def change_point_color(point_id, new_color):
    try:
        data = {"point_id": point_id, "new_color": new_color}
        response = requests.post(f"{base_url()}/change_point_color", json=data, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def set_point_position(point_id, positon):
    try:
        data = {"point_id": point_id, "position": position}
        response = requests.post(f"{base_url()}/set_point_position", json=data, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def create_joint(point1_id, point2_id):
    try:
        data = {"point1_id": point1_id, "point2_id": point2_id}
        response = requests.post(f"{base_url()}/create_joint", json=data, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def change_joint(joint_id, point1_id, point2_id):
    try:
        data = {"joint_id": joint_id, "point1_id": point1_id, "point2_id": point2_id}
        response = requests.post(f"{base_url()}/change_joint", json=data, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

def delete_joint(joint_id):
    try:
        response = requests.post(f"{base_url()}/delete_joint", json={"joint_id": joint_id}, auth = auth())
        return response.json()
    except Exception:
        return {"error": True, "value": "Failed to make request", "stack": traceback.format_exc()}

File: Sewing/Tools/Annotation/Labeling/set_up_routes/data_routes/backend/points_and_joints.js
const pointJointActions = require("../data_actions/points_and_joints.js");

module.exports = (app) => {
    app.get("/get_points_for_project", async (req, res) => {
        try {
            const result = await pointJointActions.getPointsForProject(req.body.project_id);
            return res.json({
                error: false,
                points: result
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.toString(),
                stack: err.stack
            });
        }
    });

    app.get("/get_joints_for_project", async (req, res) => {
        try {
            const result = await pointJointActions.getJointsForProject(req.body.project_id);
            return res.json({
                error: false,
                joints: result
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.toString(),
                stack: err.stack
            });
        }
    });

    app.post(`/create_point`, async (req, res) => {
        try {
            const result = await pointJointActions.createPoint(req.body.project_id, req.body.name, req.body.color);
            return res.json({
                error: false,
                ...result
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.message,
                stack: err.stack
            });
        }
    });

    app.post(`/rename_point`, async (req, res) => {
        try {
            const result = await pointJointActions.renamePoint(req.body.point_id, req.body.new_name);
            return res.json({
                error: false,
                ...result
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.message,
                stack: err.stack
            });
        }
    });

    app.post(`/set_point_position`, async (req, res) => {
        try {
            const result = await pointJointActions.changePointPositon(req.body.point_id, req.body.position);
            return res.json({
                error: false,
                ...result
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.message,
                stack: err.stack
            });
        }
    });

    app.post(`/delete_point`, async (req, res) => {
        try {
            await pointJointActions.deletePoint(req.body.point_id);
            return res.json({
                error: false,
                message: "Success!"
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.toString(),
                stack: err.stack
            });
        }
    });

    app.post(`/change_point_color`, async (req, res) => {
        try {
            const result = await pointJointActions.changePointColor(req.body.point_id, req.body.new_color);
            return res.json({
                error: false,
                ...result
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.toString(),
                stack: err.stack
            });
        }
    });

    app.post(`/create_joint`, async (req, res) => {
        try {
            const result = await pointJointActions.createJoint(req.body.point1_id, req.body.point2_id);
            return res.json({
                error: false,
                joint_id: result
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.message,
                stack: err.stack
            });
        }
    });

    app.post(`/change_joint`, async (req, res) => {
        try {
            const result = await pointJointActions.changeJoint(req.body.joint_id, req.body.point1_id, req.body.point2_id);
            return res.json({
                error: false,
                ...result
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.message,
                stack: err.stack
            });
        }
    });

    app.post(`/delete_joint`, async (req, res) => {
        try {
            await pointJointActions.deleteJoint(req.body.joint_id);
            return res.json({
                error: false,
                message: "Success!"
            });
        } catch (err) {
            return res.json({
                error: true,
                message: err.toString(),
                stack: err.stack
            });
        }
    });
}
